<table align="center" width=100%>
    <tr>
        <td width="15%">
            <img src="in_class.png">
        </td>
        <td>
            <div align="center">
                <font color="#21618C" size=8px>
                    <b> Inclass - Lab <br>(Day 3)
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

### About the data set  (Life Expectancy data)

The dataset contains information about several health and economic factors that contribute to average life expectancy for different countries. <br> Attribute information: 

**Country:** Name of the country

**Year:** Year the observations were recorded

**Status:** Whether the country is Developed or Developing

**Adult_Mortality:** Mortality rate for age group 15-60 out of every 1000 individuals of the population

**Infant_Deaths:** Number of infant deaths per 1000 population

**Alcohol:** Alcohol, recorded per capita (15+) consumption (in litres of pure alcohol)

**Hepatitis B:** Hepatitis B (HepB) immunization coverage for 1 year olds (Percentage)

**Measles:** Number of reported cases for measles per 1000 from population

**BMI:** Average Body Mass Index for entire population

**Underfive_Deaths:** Number of deaths under 5 years of age per 1000 population

**Polio:** Polio (Pol3) immunization coverage for 1 year olds (Percentage)

**Diphtheria:** Diphtheria tetanus toxoid and pertussis (DTP3) immunization coverage for 1 year olds (Percentage)

**HIV:** Deaths per 1000 live births due to HIV/AIDS (0-4 years)

**GDP:** Gross Domestic Product per capita (in USD)

**Population:** Population of the country

**Malnourished10_19:** Prevalence of malnutrition among children and adolescents for Age 10 to 19 (Percentage)

**Malnourished5_9:** Prevalence of malnutrition among children for Age 5 to 9 (Percentage)

**Income_Index:** Human Development Index (HDI) in terms of national income per capita (index ranging from 0 to 1)

**Schooling:** Number of years of Schooling

**Life_Expectancy:** Life Expectancy in age for the country


##  Table of Content

1. **[Data Preparation](#eda)**
2. **[Linear Regression](#lf)**
3. **[Gradient Descent](#gd)**
4. **[Regularization](#reg)**
5. **[Grid Search](#gs)**

**Import the required libraries**

In [1]:
# type your code here

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

#### Load the Life expectancy dataset and display the first five records 

In [2]:
# load the csv file

# type your code here
df=pd.read_csv("Life_Expectancy.csv",index_col=0)
df=df.sample(frac=1,random_state=0)
# display the first five observations

# type your code here
df.head()

Year      Status  Adult_Mortality  Infant_Deaths  Alcohol  \
Country                                                                 
Madagascar  2015  Developing             22.0             28      NaN   
Bhutan      2015  Developing            211.0              0      NaN   
Chile       2015  Developing             82.0              2      NaN   
Mali        2015  Developing            266.0             52      NaN   
Zimbabwe    2015  Developing            336.0             22      NaN   

            Hepatitis B  Measles    BMI  Underfive_Deaths   Polio  \
Country                                                             
Madagascar         69.0         3   2.5                 38     71   
Bhutan             99.0        11  24.5                  0     98   
Chile              97.0         9  63.8                  2     96   
Mali               64.0       215  23.8                 84     66   
Zimbabwe           87.0         0  31.8                 32     88   

            Diphtheria   HIV           GDP  Population  Malnourished10_19  \
Country                                                                     
Madagascar           69  0.3     41.857595   2423488.0                7.1   
Bhutan               99  0.5   2613.645177    787386.0               15.4   
Chile                96  0.1  13653.226400  17762681.0                0.8   
Mali                 64  1.5    729.725338   1746795.0                7.7   
Zimbabwe             87  6.2    118.693830  15777451.0                5.6   

            Malnourished5_9  Income_Index  Schooling  Life_Expectancy  
Country                                                                
Madagascar              7.1         0.511       10.3             65.5  
Bhutan                 16.0         0.604       12.5             69.8  
Chile                   0.8         0.845       16.3             85.0  
Mali                    7.5         0.438        8.4             58.2  
Zimbabwe                5.5         0.507       10.3             67.0

#### Check the shape of the data and display its information

In [3]:
# check the size of the dataframe

# type your code here
df.shape

(182, 19)

In [4]:
# display dataset information understand the dataset

# type your code here
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 182 entries, Madagascar to Great Britain
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Year               182 non-null    int64  
 1   Status             182 non-null    object 
 2   Adult_Mortality    173 non-null    float64
 3   Infant_Deaths      182 non-null    int64  
 4   Alcohol            15 non-null     float64
 5   Hepatitis B        173 non-null    float64
 6   Measles            182 non-null    int64  
 7   BMI                178 non-null    float64
 8   Underfive_Deaths   182 non-null    int64  
 9   Polio              182 non-null    int64  
 10  Diphtheria         182 non-null    int64  
 11  HIV                182 non-null    float64
 12  GDP                157 non-null    float64
 13  Population         141 non-null    float64
 14  Malnourished10_19  178 non-null    float64
 15  Malnourished5_9    178 non-null    float64
 16  Income_Index

### Let's begin with some hands-on practice exercises

<a id="eda"> </a>
## 1. Data Preparation

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>1. The dataset gives the life expectancy for different countries. Check if any information about life expectancy is missing from the given records. How do we fix the problem if present?</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [5]:
# type your code here
df.isnull().sum()

Year                   0
Status                 0
Adult_Mortality        9
Infant_Deaths          0
Alcohol              167
Hepatitis B            9
Measles                0
BMI                    4
Underfive_Deaths       0
Polio                  0
Diphtheria             0
HIV                    0
GDP                   25
Population            41
Malnourished10_19      4
Malnourished5_9        4
Income_Index          13
Schooling              9
Life_Expectancy        9
dtype: int64

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>2. How do we handle the missing values from the entire dataset?</b>
                </font>
            </div>
        </td>
    </tr>
</table>

The alcohol column has over 80 % nulls, hence it would be better to drop the column.

as for the other columns we can use ffill to fill in the nulls, which will not change the mean and distribution of the data.

In [6]:
df.drop(columns='Alcohol',inplace =True)
df.fillna(method='ffill',inplace=True)

In [7]:
df.isnull().sum()

Year                 0
Status               0
Adult_Mortality      0
Infant_Deaths        0
Hepatitis B          0
Measles              0
BMI                  0
Underfive_Deaths     0
Polio                0
Diphtheria           0
HIV                  0
GDP                  0
Population           0
Malnourished10_19    0
Malnourished5_9      0
Income_Index         0
Schooling            0
Life_Expectancy      0
dtype: int64

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>3. Are there any redundant features in the data?</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [8]:
# type your code here


# The alcohol column with 80% nulls can be considered redundant
# No duplicates are found

df[df.duplicated()==True]

Empty DataFrame
Columns: [Year, Status, Adult_Mortality, Infant_Deaths, Hepatitis B, Measles , BMI, Underfive_Deaths , Polio, Diphtheria , HIV, GDP, Population, Malnourished10_19, Malnourished5_9, Income_Index, Schooling, Life_Expectancy]
Index: []

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>4. Perform dummy encoding for appropriate variables of the dataset if required</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [9]:
# type your code here
df.head()

Year      Status  Adult_Mortality  Infant_Deaths  Hepatitis B  \
Country                                                                     
Madagascar  2015  Developing             22.0             28         69.0   
Bhutan      2015  Developing            211.0              0         99.0   
Chile       2015  Developing             82.0              2         97.0   
Mali        2015  Developing            266.0             52         64.0   
Zimbabwe    2015  Developing            336.0             22         87.0   

            Measles    BMI  Underfive_Deaths   Polio  Diphtheria   HIV  \
Country                                                                  
Madagascar         3   2.5                 38     71           69  0.3   
Bhutan            11  24.5                  0     98           99  0.5   
Chile              9  63.8                  2     96           96  0.1   
Mali             215  23.8                 84     66           64  1.5   
Zimbabwe           0  31.8                 32     88           87  6.2   

                     GDP  Population  Malnourished10_19  Malnourished5_9  \
Country                                                                    
Madagascar     41.857595   2423488.0                7.1              7.1   
Bhutan       2613.645177    787386.0               15.4             16.0   
Chile       13653.226400  17762681.0                0.8              0.8   
Mali          729.725338   1746795.0                7.7              7.5   
Zimbabwe      118.693830  15777451.0                5.6              5.5   

            Income_Index  Schooling  Life_Expectancy  
Country                                               
Madagascar         0.511       10.3             65.5  
Bhutan             0.604       12.5             69.8  
Chile              0.845       16.3             85.0  
Mali               0.438        8.4             58.2  
Zimbabwe           0.507       10.3             67.0

In [10]:
df = pd.get_dummies(df,columns=['Status'], drop_first=True)
df.head()

Year  Adult_Mortality  Infant_Deaths  Hepatitis B  Measles    BMI  \
Country                                                                         
Madagascar  2015             22.0             28         69.0         3   2.5   
Bhutan      2015            211.0              0         99.0        11  24.5   
Chile       2015             82.0              2         97.0         9  63.8   
Mali        2015            266.0             52         64.0       215  23.8   
Zimbabwe    2015            336.0             22         87.0         0  31.8   

            Underfive_Deaths   Polio  Diphtheria   HIV           GDP  \
Country                                                                
Madagascar                 38     71           69  0.3     41.857595   
Bhutan                      0     98           99  0.5   2613.645177   
Chile                       2     96           96  0.1  13653.226400   
Mali                       84     66           64  1.5    729.725338   
Zimbabwe                   32     88           87  6.2    118.693830   

            Population  Malnourished10_19  Malnourished5_9  Income_Index  \
Country                                                                    
Madagascar   2423488.0                7.1              7.1         0.511   
Bhutan        787386.0               15.4             16.0         0.604   
Chile       17762681.0                0.8              0.8         0.845   
Mali         1746795.0                7.7              7.5         0.438   
Zimbabwe    15777451.0                5.6              5.5         0.507   

            Schooling  Life_Expectancy  Status_Developing  
Country                                                    
Madagascar       10.3             65.5                  1  
Bhutan           12.5             69.8                  1  
Chile            16.3             85.0                  1  
Mali              8.4             58.2                  1  
Zimbabwe         10.3             67.0                  1

<a id="lf"> </a>
## 2. Linear Regression

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>5.  Build a full model on the given data. Check whether the obtained residuals have constant variance
</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [11]:
# type your code here

import statsmodels.api as sm
from sklearn.model_selection import train_test_split

y=df['Life_Expectancy']
x=df.drop('Life_Expectancy',axis=1)
xc=sm.add_constant(x)

model = sm.OLS(y,xc).fit()

ypred = model.predict(xc)

residuals = model.resid

In [12]:
# to check for variance of the residuals, we have to use goldfeld test

# H0: variance is constant accross the residual
# HA: variance is not constant accross the range of data

from statsmodels.stats.api import het_goldfeldquandt

het_goldfeldquandt(y=residuals,x=xc)

(0.7818091055685484, 0.8523630155288722, 'increasing')

### Conclusion:
The pvalue is  0.852, which is greater than 0.05, hence we accept H0, ie; the variance is constant across the residual

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>6. Build a model to study the impact of immunization on life expectancy, using least squares regression. Find the unexplained variation for the model
</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [13]:
# type your code here
y=df['Life_Expectancy']
x=df[['Polio','Hepatitis B','Diphtheria ']]
xc=sm.add_constant(x)

model = sm.OLS(y,xc).fit()

ypred = model.predict(xc)

residuals=model.resid

In [14]:
# unsexplained variation of the model is the SSE or sum of squares of residuals

sse = np.sum(residuals**2)
print('The unexplained variation in the model = ',sse)

The unexplained variation in the model =  8086.042918989142


<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>7. Build a model to study the impact of malnutrition on life expectancy, using least squares regression. Interpret the coeffients</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [15]:
# type your code here

y=df['Life_Expectancy']
x=df[['Malnourished10_19','Malnourished5_9']]
xc=sm.add_constant(x)

model = sm.OLS(y,xc).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        Life_Expectancy   R-squared:                       0.214
Model:                            OLS   Adj. R-squared:                  0.206
Method:                 Least Squares   F-statistic:                     24.43
Date:                Sun, 03 Jan 2021   Prob (F-statistic):           4.14e-10
Time:                        09:36:27   Log-Likelihood:                -615.97
No. Observations:                 182   AIC:                             1238.
Df Residuals:                     179   BIC:                             1248.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                76.0613      0.778     97.721      0.000      74.525      77.597
Malnourished10_19    -0.3954      0.570     -0.694      0.488      -1.519       0.728
Malnourished5_9      -0.5087      0.558     -0.912      0.363      -1.610       0.592
==============================================================================
Omnibus:                        7.731   Durbin-Watson:                   1.853
Prob(Omnibus):                  0.021   Jarque-Bera (JB):                8.102
Skew:                          -0.508   Prob(JB):                       0.0174
Kurtosis:                       2.810   Cond. No.                         12.8
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Conclusion:
Coeff for Malnurished10_19:
    
    With every 1% increase in Malnurished10_19, the Life Expectency of the country reduces by 0.3954 years.
    
Coeff for Malnurished10_19:
    
    With every 1% increase in Malnurished5_9, the Life Expectency of the country reduces by 0.5087 years.

<a id="gd"> </a>
## 3. Gradient Descent

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>8. Build a model to study the influence of adult mortality rate on life expectancy using both the least squares regression and gradient descent approach. Are the fits for both the models different?</b>
                </font>
            </div>
        </td>
    </tr>
</table>

### OLS Method:

In [16]:
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_squared_error

y=df['Life_Expectancy']
x=df[['Adult_Mortality']]
xc=sm.add_constant(x)

xtrain,xtest,ytrain,ytest = train_test_split(xc,y,test_size=0.3,random_state=1)

model = sm.OLS(ytrain,xtrain).fit()

ypredtest_ols = model.predict(xtest)
ypredtrain_ols = model.predict(xtrain)

r2_ols_test = r2_score(ytest,ypredtest_ols)
r2_ols_train  = r2_score(ytrain,ypredtrain_ols)

rmse_ols_test = np.sqrt(mean_squared_error(ytest,ypredtest_ols))
rmse_ols_train = np.sqrt(mean_squared_error(ytrain,ypredtrain_ols))

### Linear Regression  (Gradient Descent)

In [17]:
from sklearn.linear_model import LinearRegression

y=df['Life_Expectancy']
x=df[['Adult_Mortality']]

xtrain,xtest,ytrain,ytest = train_test_split(xc,y,test_size=0.3,random_state=1)

lr= LinearRegression()

lr.fit(xtrain,ytrain)

ypredtest_lr = lr.predict(xtest)
ypredtrain_lr = lr.predict(xtrain)

r2_lr_test = r2_score(ytest,ypredtest_ols)
r2_lr_train  = r2_score(ytrain,ypredtrain_ols)

rmse_lr_test = np.sqrt(mean_squared_error(ytest,ypredtest_lr))
rmse_lr_train = np.sqrt(mean_squared_error(ytrain,ypredtrain_lr))

In [18]:
print('OLS: ')
print('R2 for train data : ',r2_ols_train)
print('R2 for test data  : ',r2_ols_test)
print('\n')
print('Linear Regression (Gradient Descent): ')
print('R2 for train data : ',r2_lr_train)
print('R2 for test data  : ',r2_lr_test)
print('\n')
print('OLS :')
print('RmSE for train data : ',rmse_ols_train)
print('RMSE for test data  : ',rmse_ols_test)
print('\n')
print('Linear Regression (Gradient Descent) :')
print('RMSE for train data : ',rmse_lr_train)
print('RMSE for test data  : ',rmse_lr_test)

OLS: 
R2 for train data :  0.7665783826247218
R2 for test data  :  0.1816655264987097


Linear Regression (Gradient Descent): 
R2 for train data :  0.7665783826247218
R2 for test data  :  0.1816655264987097


OLS :
RmSE for train data :  3.844031814615605
RMSE for test data  :  7.453680454377165


Linear Regression (Gradient Descent) :
RMSE for train data :  3.844031814615605
RMSE for test data  :  7.4536804543771735


## Conclusion: 

We observe that the RMSE and R2 values for test and train data for both OLS and Gradient Descent method are equal in value.

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>9. Build a model to analyze the influence of immunization on life expectancy. Use SGD and plot a horizontal multiple barchart to compare the values of beta coefficients with values obtained by the full model build least squares regression.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [19]:
# type your code here


<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>10. Build a full model using least squares regression. Check whether the model overfits the training data or not.
</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [20]:
# type your code here

y=df['Life_Expectancy']
x=df.drop('Life_Expectancy',axis=1)
xc=sm.add_constant(x)

xtrain,xtest,ytrain,ytest = train_test_split(xc,y,test_size=0.3,random_state=1)

model = sm.OLS(ytrain,xtrain).fit()

ypredtest_ols = model.predict(xtest)
ypredtrain_ols = model.predict(xtrain)

r2_ols_test = r2_score(ytest,ypredtest_ols)
r2_ols_train  = r2_score(ytrain,ypredtrain_ols)

rmse_ols_test = np.sqrt(mean_squared_error(ytest,ypredtest_ols))
rmse_ols_train = np.sqrt(mean_squared_error(ytrain,ypredtrain_ols))

print('R2 for train data : ',r2_ols_train)
print('R2 for test data  : ',r2_ols_test,'\n')
print('RmSE for train data : ',rmse_ols_train)
print('RMSE for test data  : ',rmse_ols_test)

R2 for train data :  0.8974827547365788
R2 for test data  :  0.7821794018539554 

RmSE for train data :  2.5475041507622502
RMSE for test data  :  3.8455177471784547


### Conclusion:

When we compre the RMSE and R2 of the test and train part of the data with the model we can't see stark discrepancy between the two.
Hence we can conclude that the model isn't overfitting.

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>11. Can we use a linear regression model to analyze how all features from the dataset impact life expectancy?</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [21]:
# type your code here

x=df.drop(columns='Life_Expectancy')
y=df['Life_Expectancy']
xc= sm.add_constant(x)

model = sm.OLS(y,xc).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        Life_Expectancy   R-squared:                       0.880
Model:                            OLS   Adj. R-squared:                  0.868
Method:                 Least Squares   F-statistic:                     70.81
Date:                Sun, 03 Jan 2021   Prob (F-statistic):           2.66e-66
Time:                        09:36:28   Log-Likelihood:                -444.92
No. Observations:                 182   AIC:                             925.8
Df Residuals:                     164   BIC:                             983.5
Df Model:                          17                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const              1617.4571   1113.508      1.453      0.148    -581.203    3816.117
Year                 -0.7751      0.553     -1.403      0.163      -1.866       0.316
Adult_Mortality      -0.0233      0.003     -6.693      0.000      -0.030      -0.016
Infant_Deaths         0.0955      0.034      2.827      0.005       0.029       0.162
Hepatitis B           0.0250      0.018      1.364      0.174      -0.011       0.061
Measles           -7.316e-05   5.73e-05     -1.276      0.204      -0.000       4e-05
BMI                  -0.0080      0.013     -0.592      0.555      -0.035       0.019
Underfive_Deaths     -0.0698      0.024     -2.861      0.005      -0.118      -0.022
Polio                 0.0133      0.013      1.024      0.307      -0.012       0.039
Diphtheria            0.0072      0.023      0.313      0.754      -0.038       0.052
HIV                  -0.5409      0.226     -2.392      0.018      -0.987      -0.094
GDP               -9.929e-06   2.19e-05     -0.454      0.651   -5.31e-05    3.33e-05
Population         3.146e-09   9.56e-09      0.329      0.743   -1.57e-08     2.2e-08
Malnourished10_19    -0.1148      0.247     -0.465      0.642      -0.602       0.372
Malnourished5_9      -0.0781      0.244     -0.320      0.750      -0.560       0.404
Income_Index         30.3752      4.239      7.166      0.000      22.005      38.745
Schooling            -0.1615      0.167     -0.966      0.335      -0.491       0.169
Status_Developing    -1.4626      0.703     -2.080      0.039      -2.851      -0.074
==============================================================================
Omnibus:                        3.806   Durbin-Watson:                   1.930
Prob(Omnibus):                  0.149   Jarque-Bera (JB):                4.644
Skew:                          -0.029   Prob(JB):                       0.0981
Kurtosis:                       3.780   Cond. No.                     1.48e+11
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.48e+11. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### Conclusions:
#### The above dataframe shows the coeffiecients of all the independent variables in the model.
#### The negative sign signifies negative correlation with the dependent variable, ie; with the increase in these variables the life expectancy reduces. for eg: with unit increase in adult mortality rate, the life expectancy reduces by 0.02326 years
#### The positive sign signifies positive correlation with the dependent variable, ie; with the increas in these variables the life expectancy increases. for eg: for every unit increase in Income_index, the life expectancy increases by 30.3 years
#### But there are several features that do not contribute to the model which can be checked using the pvalue >0.05.

<a id="reg"> </a>
## 4. Regularization

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>12. Can we perform regression analysis without eliminating features involved in multicollinearity detected in question 11?</b>
                </font>
            </div>
        </td>
    </tr>

</table>

In [34]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split


ss= StandardScaler()

x = df.drop(columns='Life_Expectancy')
y = df['Life_Expectancy']

xtrain,xtest,ytrain,ytest = train_test_split(x,y,test_size = 0.3, random_state = 0)

x_ss = ss.fit_transform(x)
xtrain_ss = ss.fit_transform(xtrain)
xtest_ss  = ss.transform(xtest)

In [47]:
# type your code here
from sklearn.linear_model import Ridge, RidgeCV
from sklearn.metrics import r2_score

alphas = np.logspace(-3,0,500)

r_cv = RidgeCV(alphas = alphas,cv = 5)
r_cv.fit(x_ss,y)

RidgeCV(alphas=array([0.001     , 0.00101394, 0.00102807, 0.0010424 , 0.00105693,
       0.00107167, 0.00108661, 0.00110175, 0.00111711, 0.00113268,
       0.00114847, 0.00116448, 0.00118071, 0.00119717, 0.00121386,
       0.00123078, 0.00124794, 0.00126533, 0.00128297, 0.00130085,
       0.00131899, 0.00133737, 0.00135602, 0.00137492, 0.00139408,
       0.00141352, 0.00143322, 0.0014532 , 0.00147345, 0.00149399,
       0.001...
       0.71731749, 0.72731651, 0.73745491, 0.74773463, 0.75815765,
       0.76872595, 0.77944158, 0.79030657, 0.80132301, 0.81249302,
       0.82381873, 0.83530232, 0.84694598, 0.85875195, 0.87072248,
       0.88285988, 0.89516647, 0.90764461, 0.92029668, 0.93312512,
       0.94613238, 0.95932095, 0.97269336, 0.98625218, 1.        ]),
        cv=5, fit_intercept=True, gcv_mode=None, normalize=False, scoring=None,
        store_cv_values=False)

In [48]:
r_cv.alpha_

0.21511234301821672

In [50]:
ridge = Ridge(alpha=r_cv.alpha_)

ridge.fit(xtrain_ss,ytrain)

ypredtrain_r = ridge.predict(xtrain_ss)
ypredtest_r  = ridge.predict(xtest_ss)

r2_train_r = r2_score(ytrain,ypredtrain_r)
r2_test_r  = r2_score(ytest, ypredtest_r)

print('Train r2 : ',r2_train_r)
print('Test r2  : ',r2_test_r)

Train r2 :  0.8975458177306752
Test r2  :  0.8144838924080635


<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>13. Build a model to study the impact of diseases and immunization on life expectancy. Identify the significant variables from the model. Use 'alpha = 1.5' if required.
</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [59]:
# type your code here
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Lasso

ss = StandardScaler()

x  = df[['Hepatitis B','Measles ','Polio','Diphtheria ','HIV']]
y  = df['Life_Expectancy']

xtrain,xtest,ytrain,ytest = train_test_split(x,y,test_size=0.3,random_state=0)

x_ss      = ss.fit_transform(x)
xtrain_ss = ss.fit_transform(xtrain)
xtest_ss  = ss.transform(xtest)

lasso = Lasso(alpha = 1.5)

lasso.fit(xtrain_ss,ytrain)

pd.DataFrame(lasso.coef_,index=x.columns,columns= ['Lasso_coeffs'])

Lasso_coeffs
Hepatitis B      0.000000
Measles         -0.000000
Polio            1.247326
Diphtheria       0.924996
HIV             -2.612268

## Conclusion:
#### From the above dataframe we can conclude that the features with non zero coefficients are significant, ie; Polio, Diphtheria, HIV

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>14. Build a model to study the impact of malnutrition and diseases on life expectancy. Identify the insignificant variables from the model. Use 'alpha = 1' if required.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [61]:
# type your code here
x = df[['Measles ','HIV','Malnourished10_19','Malnourished5_9']]
y = df['Life_Expectancy']

xtrain, xtest, ytrain,ytest = train_test_split(x,y,test_size = 0.3, random_state = 0)

x_ss = ss.fit_transform(x)
xtrain_ss = ss.fit_transform(xtrain)
xtest_ss  = ss.transform(xtest)

lasso = Lasso(alpha = 1)

lasso.fit(xtrain_ss,ytrain)
pd.DataFrame(lasso.coef_,index= x.columns, columns = ['Lasso Coeffs'])

Lasso Coeffs
Measles               -0.000000
HIV                   -3.660492
Malnourished10_19     -0.000000
Malnourished5_9       -2.170585

## Conclusions:
#### From the dataframe, the variables with non zero coeffs are the significant variables, ie; HIV, Malnourished5_9

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>15. Build a full model using the elastic net regression. Use 'alpha = 0.1' and 'l1_ratio=0.001'. Also, compute the RMSE for train and test sets.
 </b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [63]:
# type your code here
from sklearn.linear_model import ElasticNet
from sklearn.metrics import mean_squared_error

y = df['Life_Expectancy']
x = df.drop(columns = 'Life_Expectancy')

xtrain, xtest, ytrain, ytest = train_test_split(x,y, test_size = 0.3, random_state = 0)

xtrain_ss =  ss.fit_transform(xtrain)
xtest_ss  =  ss.transform(xtest)

enet = ElasticNet(alpha=0.1,l1_ratio=0.001)
enet.fit(xtrain_ss,ytrain)

ypred_train_e = enet.predict(xtrain_ss)
ypred_test_e  = enet.predict(xtest_ss)

rmse_train = mean_squared_error(ytrain,ypred_train_e)
rmse_test  = mean_squared_error(ytest,ypred_test_e)

print('RMSE for train Elastic Net Model = ',rmse_train)
print('RMSE for test Elastic Net Model  = ',rmse_test)

RMSE for train Elastic Net Model =  6.967148370620074
RMSE for test Elastic Net Model  =  13.188187662874578


Model is over fitting as rmse for test set is almost double that of train set

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>16. Build a full model. Identify the features that are significantly influencing the life expectancy. Use 'alpha = 0.5' if required. </b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [69]:
# type your code here

from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Lasso

ss = StandardScaler()

x  = df.drop(columns='Life_Expectancy')
y  = df['Life_Expectancy']

xtrain,xtest,ytrain,ytest = train_test_split(x,y,test_size=0.3,random_state=0)

x_ss      = ss.fit_transform(x)
xtrain_ss = ss.fit_transform(xtrain)
xtest_ss  = ss.transform(xtest)

lasso = Lasso(alpha = 0.5)

lasso.fit(xtrain_ss,ytrain)

cf=pd.DataFrame(lasso.coef_,index=x.columns,columns= ['Lasso_coeffs'])

print('The significant variables with non zero coeffs are :')
cf[np.abs(cf['Lasso_coeffs'])>0.000000]

The significant variables with non zero coeffs are :


Lasso_coeffs
Adult_Mortality       -2.127882
Hepatitis B            0.128865
Polio                  0.152843
Diphtheria             0.561193
HIV                   -0.859409
Malnourished10_19     -0.319802
Income_Index           3.991726
Status_Developing     -0.186406

<a id="gs"> </a>
## 5. Grid Search 

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>17. If ridge regression is used to build a full model. Which value is the best alpha from the list of values?<br><br>Consider the alpha values: [7,8,9,10,11]</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [71]:
# type your code here
from sklearn.linear_model import RidgeCV

x  = df.drop(columns='Life_Expectancy')
y  = df['Life_Expectancy']

ss = StandardScaler()

x_ss      = ss.fit_transform(x)

ridgecv = RidgeCV(alphas=[7,8,9,10,11],cv=5,scoring = 'r2')

ridgecv.fit(x_ss,y)

RidgeCV(alphas=array([ 7,  8,  9, 10, 11]), cv=5, fit_intercept=True,
        gcv_mode=None, normalize=False, scoring='r2', store_cv_values=False)

In [78]:
print('The best value of alpha for Ridge is : ',ridgecv.alpha_)

The best value of alpha for Ridge is :  11


<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>18. If lasso regression is used to build a full model. Which value is the best alpha from the list of values?<br><br>Consider the alpha values: [7,8,9,10,11]</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [79]:
# type your code here
from sklearn.linear_model import LassoCV

x  = df.drop(columns='Life_Expectancy')
y  = df['Life_Expectancy']

ss = StandardScaler()

x_ss      = ss.fit_transform(x)

lasscv =LassoCV(alphas= [7,8,9,10,11],cv= 5)

lasscv.fit(x_ss,y)

LassoCV(alphas=[7, 8, 9, 10, 11], copy_X=True, cv=5, eps=0.001,
        fit_intercept=True, max_iter=1000, n_alphas=100, n_jobs=None,
        normalize=False, positive=False, precompute='auto', random_state=None,
        selection='cyclic', tol=0.0001, verbose=False)

In [80]:
print('The best alpha for Lasso is : ',lasscv.alpha_)

The best alpha for Lasso is :  7


<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>19. Perform regression analysis to study the impact of malnutrition and diseases on life expectancy. Determine the optimal value of alpha and mixing parameter if elastic net regression is used to build the model. Consider the following list of values:<br><br> alpha : [0.1,0.5,1,1.5,2,2.5,3,3.5,4]<br> l1_ratio : [0.5,0.55,0.6,0.65,0.7,0.75,0.8,0.85,0.9] </b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [87]:
# type your code here
from sklearn.linear_model import ElasticNetCV

x = df[['Measles ','HIV','Malnourished10_19','Malnourished5_9']]
y = df['Life_Expectancy']

xtrain, xtest, ytrain,ytest = train_test_split(x,y,test_size = 0.3, random_state = 0)

x_ss = ss.fit_transform(x)
xtrain_ss = ss.fit_transform(xtrain)
xtest_ss  = ss.transform(xtest)

enetcv=ElasticNetCV(alphas=[0.1,0.5,1,1.5,2,2.5,3,3.5,4],l1_ratio=[0.5,0.55,0.6,0.65,0.7,0.75,0.8,0.85,0.9])
enetcv.fit(x_ss,y)

ElasticNetCV(alphas=[0.1, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4], copy_X=True, cv=None,
             eps=0.001, fit_intercept=True,
             l1_ratio=[0.5, 0.55, 0.6, 0.65, 0.7, 0.75, 0.8, 0.85, 0.9],
             max_iter=1000, n_alphas=100, n_jobs=None, normalize=False,
             positive=False, precompute='auto', random_state=None,
             selection='cyclic', tol=0.0001, verbose=0)

In [89]:
print('The best alpha is     : ',enetcv.alpha_)
print('The best l1_ratio is  : ',enetcv.l1_ratio_)

The best alpha is     :  0.5
The best l1_ratio is  :  0.7


In [90]:
enet = ElasticNet(alpha=enetcv.alpha_,l1_ratio=enetcv.l1_ratio_)

enet.fit(xtrain_ss,ytrain)

ypred_train_e = enet.predict(xtrain_ss)
ypred_test_e  = enet.predict(xtest_ss)

print('R2 value for train ENET : ',r2_score(ytrain,ypred_train_e))
print('R2 value for test ENET  : ',r2_score(ytest, ypred_test_e))

R2 value for train ENET :  0.5319898587669215
R2 value for test ENET  :  0.34600399003396964


<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>20. Build a full model. Select the optimal value for elastic net mixing parameter if the alpha value is 0.8<br><br>Consider the values: l1_ratio : [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1]</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [93]:
# type your code here

x  = df.drop(columns='Life_Expectancy')
y  = df['Life_Expectancy']

ss = StandardScaler()

x_ss      = ss.fit_transform(x)
xtrain_ss = ss.fit_transform(xtrain)
xtest_ss  = ss.transform(xtest)

enetcv = ElasticNetCV(alphas = [0.8],l1_ratio=[0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1])

enetcv.fit(x_ss,y)

print('The best alpha is     : ',enetcv.alpha_)
print('The best l1_ratio is  : ',enetcv.l1_ratio_)

The best alpha is     :  0.8
The best l1_ratio is  :  1.0


In [94]:
enet= ElasticNet(alpha = 0.8,l1_ratio=1)

enet.fit(xtrain_ss,ytrain)

ypred_train_e = enet.predict(xtrain_ss)
ypred_test_e  = enet.predict(xtest_ss)

print('R2 value for train ENET : ',r2_score(ytrain,ypred_train_e))
print('R2 value for test ENET  : ',r2_score(ytest, ypred_test_e))

R2 value for train ENET :  0.5290637674375633
R2 value for test ENET  :  0.35848557958744165
